ir/src/symbol_table/mod.rs

```diff
@@ -210,28 +206,49 @@ impl SymbolTable {
         }
     }
 
-    /// Looks up a [TraceBindingAccess] by its identifier name and returns an equivalent
-    /// [TraceAccess].
+    /// Looks up a [SymbolAccess] by its identifier name and returns an equivalent [TraceAccess] if
+    /// the symbol access references a single value in the trace.
     ///
     /// # Errors
     /// Returns an error if:
     /// - the identifier was not in the symbol table.
-    /// - the identifier was not declared as a trace column binding.
-    /// TODO: update docs
-    pub(crate) fn get_trace_binding_access(
+    /// - the identifier was not declared as a trace binding.
+    /// - the access would not return a single value in the execution trace.
+    pub(crate) fn get_trace_access(
         &self,
-        trace_access: &TraceBindingAccess,
+        symbol_access: &SymbolAccess,
     ) -> Result<TraceAccess, SemanticError> {
-        let symbol = self.get_symbol(trace_access.name())?;
-        trace_access.validate(symbol)?;
-
-        let SymbolBinding::Trace(columns) = symbol.binding() else { unreachable!("validation of named trace access failed.") };
-        Ok(TraceAccess::new(
-            columns.trace_segment(),
-            columns.offset() + trace_access.col_offset(),
-            1,
-            trace_access.row_offset(),
-        ))
+        let symbol = self.get_symbol(symbol_access.name())?;
+
+        match symbol.binding() {
+            SymbolBinding::Trace(columns) => {
+                let col_offset = match symbol_access.access_type() {
+                    AccessType::Default => columns.offset(),
+                    AccessType::Vector(idx) => {
+                        if *idx >= columns.size() {
+                            return Err(SemanticError::invalid_access_type(
+                                symbol,
+                                symbol_access.access_type(),
+                            ));
+                        }
+                        columns.offset() + *idx
+                    }
+                    _ => {
+                        return Err(SemanticError::invalid_access_type(
+                            symbol,
+                            symbol_access.access_type(),
+                        ));
+                    }
+                };
+                Ok(TraceAccess::new(
+                    columns.trace_segment(),
+                    col_offset,
+                    1,
+                    symbol_access.offset(),
+                ))
+            }
+            _ => Err(SemanticError::not_a_trace_column_identifier(symbol)),
+        }
```

Comment:
nit: I wonder if it would be possible to reduce nesting here a bit. Since we are interested only in a single variant, maybe we can do something like this upfront:
```Rust
let columns = match symbol.binding() {
    SymbolBinding::Trace(columns) => columns,
    _ => return Err(SemanticError::not_a_trace_column_identifier(symbol)),
};
```